smart_open/s3.py

```diff
@@ -605,31 +695,66 @@ def read(self, size=-1):
         # HTTP connection and try again.  Usually, a single retry attempt is
         # enough to recover, but we try multiple times "just in case".
         #
-        for attempt, seconds in enumerate([1, 2, 4, 8, 16], 1):
-            try:
-                if size == -1:
-                    binary = self._body.read()
-                else:
-                    binary = self._body.read(size)
-            except (
-                ConnectionResetError,
-                botocore.exceptions.BotoCoreError,
-                urllib3.exceptions.HTTPError,
-            ) as err:
-                logger.warning(
-                    '%s: caught %r while reading %d bytes, sleeping %ds before retry',
-                    self,
-                    err,
-                    size,
-                    seconds,
-                )
-                time.sleep(seconds)
+        binary_collected = bytearray()
+
+        while True:
+            if self._body is None:
+                # This is necessary for the very first read() after __init__(),
+                # or after a chunk is exhausted (if using chunk_range)
                 self._open_body()
+                is_fresh_chunk = True
             else:
+                is_fresh_chunk = False
+
+            if self._position >= self._content_length or (size >= 0 and len(binary_collected) >= size):
+                break
+
+            secondss = [1, 2, 4, 8, 16]
+            for seconds in secondss:
+                try:
+                    if size == -1:
+                        binary = self._body.read()
+                    else:
+                        binary = self._body.read(size - len(binary_collected))
+                except (
+                    ConnectionResetError,
+                    botocore.exceptions.BotoCoreError,
+                    urllib3.exceptions.HTTPError,
+                ) as err:
+                    logger.warning(
+                        '%s: caught %r while reading %d bytes, sleeping %ds before retry',
+                        self,
+                        err,
+                        size,
+                        seconds,
+                    )
+                    time.sleep(seconds)
+                    self._open_body()
+                else:
+                    break
+            else:
+                raise IOError('%s: failed to read after %d attempts' % (self, len(secondss)))
+
+            if binary:
                 self._position += len(binary)
-                return binary
+                binary_collected += binary
+            else:
+                # By construction we always try to read >= 1 bytes, so if we see this
+                # we know that we are at the end of the current _body.
+                # here just in case.
+                if self._chunk_end is None or is_fresh_chunk:
```

Comment:
both `_chunk_end` and `is_fresh_chunk` variables only exist for this check. judging by the code comments, this _might_ not be necessary at all.

maybe we can construct moto tests to prove that we can avoid this added complexity?